File: packages/crypto-bls/src/poseidon.ts
/**
 * Poseidon hash over BLS12-381 Fr (canonical, unoptimized formulation).
 *
 * Hash convention matches circomlib's `Poseidon(nInputs)`: the state is
 * [0, in_0, …, in_{t-2}] (a zero capacity element first), the permutation is
 * applied, and state[0] is returned. The S-box is x^5; full rounds S-box the whole
 * state, partial rounds S-box only state[0]; the linear layer is the MDS matrix.
 */
import { add, fr, mul, pow5 } from "./field.js";
import { paramsForT, type PoseidonParams } from "./poseidon-params.js";

/** Apply the full Poseidon permutation to a state vector of length `t`. */
export function poseidonPermute(state: bigint[], p: PoseidonParams): bigint[] {
  if (state.length !== p.t) throw new Error(`state must have length t=${p.t}`);
  const { t, rf, rp, c, m } = p;
  const rounds = rf + rp;
  let s = state.map(fr);

  for (let r = 0; r < rounds; r++) {
    // Add round constants.
    for (let i = 0; i < t; i++) s[i] = add(s[i]!, c[r * t + i]!);
    // S-box: full rounds hit every lane; partial rounds hit only lane 0.
    const isFull = r < rf / 2 || r >= rf / 2 + rp;
    if (isFull) {
      for (let i = 0; i < t; i++) s[i] = pow5(s[i]!);
    } else {
      s[0] = pow5(s[0]!);
    }
    // Mix with the MDS matrix.
    const next = new Array<bigint>(t).fill(0n);
    for (let i = 0; i < t; i++) {
      let acc = 0n;
      for (let j = 0; j < t; j++) acc = add(acc, mul(m[i]![j]!, s[j]!));
      next[i] = acc;
    }
    s = next;
  }
  return s;
}

/** Poseidon hash of `inputs` (1 ≤ inputs.length ≤ 5). Returns an Fr element. */
export function poseidon(inputs: bigint[]): bigint {
  if (inputs.length < 1) throw new Error("poseidon needs at least one input");
  const t = inputs.length + 1;
  const p = paramsForT(t);
  const state = [0n, ...inputs.map(fr)];
  return poseidonPermute(state, p)[0]!;
}

/** 2-to-1 Poseidon compression (state width t=3) — the Merkle node hash. */
export const poseidon2 = (a: bigint, b: bigint): bigint => poseidon([a, b]);

File: contracts/umbra-pool/src/poseidon.rs
//! Poseidon t=3 over BLS12-381 Fr, computed on-chain with the CAP-0059 Fr host
//! functions. Uses the exact constants shared with the circuits and the TS wallet
//! (poseidon_constants.rs), so on-chain roots match off-chain roots and circuit
//! proofs verify. This is the only hashing the contract does (Tornado-style
//! on-chain incremental tree).

use soroban_sdk::crypto::bls12_381::{Bls12_381, Fr};
use soroban_sdk::{BytesN, Env};

use crate::poseidon_constants::{POSEIDON_C, POSEIDON_M, POSEIDON_RF, POSEIDON_RP};

const T: usize = 3;

pub fn fr_from_bytes(b: &BytesN<32>) -> Fr {
    Fr::from_bytes(b.clone())
}

pub fn fr_to_bytes(fr: &Fr) -> BytesN<32> {
    fr.to_bytes()
}

fn fr_const(env: &Env, b: &[u8; 32]) -> Fr {
    Fr::from_bytes(BytesN::from_array(env, b))
}

fn pow5(bls: &Bls12_381, x: &Fr) -> Fr {
    let x2 = bls.fr_mul(x, x);
    let x4 = bls.fr_mul(&x2, &x2);
    bls.fr_mul(&x4, x)
}

/// 2-to-1 Poseidon compression (the Merkle node hash and the commitment/nullifier hash).
const ROUNDS: usize = POSEIDON_RF + POSEIDON_RP;

/// Poseidon round constants + MDS matrix, deserialized to `Fr` ONCE. Building this per
/// contract call (instead of per hash) collapses ~204 `Fr::from_bytes` host calls per hash
/// — 195 round constants + 9 MDS — to a single setup. That is the dominant on-chain cost
/// when a call performs many hashes (a confidential transfer inserts two commitments =
/// 16 hashes = ~3,264 avoidable deserializations).
pub struct PoseidonParams {
    zero: Fr,
    mds: [[Fr; T]; T],
    rc: [Fr; ROUNDS * T],
}

impl PoseidonParams {
    pub fn new(env: &Env) -> Self {
        let zero = fr_const(env, &[0u8; 32]);
        let mds: [[Fr; T]; T] = [
            [fr_const(env, &POSEIDON_M[0]), fr_const(env, &POSEIDON_M[1]), fr_const(env, &POSEIDON_M[2])],
            [fr_const(env, &POSEIDON_M[3]), fr_const(env, &POSEIDON_M[4]), fr_const(env, &POSEIDON_M[5])],
            [fr_const(env, &POSEIDON_M[6]), fr_const(env, &POSEIDON_M[7]), fr_const(env, &POSEIDON_M[8])],
        ];
        // Fixed-size (no allocator needed on wasm): deserialize each round constant once.
        let rc: [Fr; ROUNDS * T] = core::array::from_fn(|i| fr_const(env, &POSEIDON_C[i]));
        Self { zero, mds, rc }
    }

    /// 2-to-1 Poseidon compression using the precomputed constants (no `from_bytes`).
    pub fn hash2(&self, env: &Env, a: &Fr, b: &Fr) -> Fr {
        let bls = env.crypto().bls12_381();
        let mut state: [Fr; T] = [self.zero.clone(), a.clone(), b.clone()];
        for r in 0..ROUNDS {
            // Add round constants.
            for i in 0..T {
                state[i] = bls.fr_add(&state[i], &self.rc[r * T + i]);
            }
            // S-box: full rounds hit every lane; partial rounds only lane 0.
            let is_full = r < POSEIDON_RF / 2 || r >= POSEIDON_RF / 2 + POSEIDON_RP;
            if is_full {
                for i in 0..T {
                    state[i] = pow5(&bls, &state[i]);
                }
            } else {
                state[0] = pow5(&bls, &state[0]);
            }
            // MDS mix.
            let mut next: [Fr; T] = [self.zero.clone(), self.zero.clone(), self.zero.clone()];
            for i in 0..T {
                let mut acc = self.zero.clone();
                for j in 0..T {
                    acc = bls.fr_add(&acc, &bls.fr_mul(&self.mds[i][j], &state[j]));
                }
                next[i] = acc;
            }
            state = next;
        }
        state[0].clone()
    }
}

/// Convenience one-shot (builds the params each call). Off the hot path — used by the
/// test oracle; hot paths build `PoseidonParams` once and call `hash2` per hash.
pub fn poseidon2(env: &Env, a: &Fr, b: &Fr) -> Fr {
    PoseidonParams::new(env).hash2(env, a, b)
}
